zebra/rt_netlink.c

```diff
@@ -2309,14 +2309,27 @@ ssize_t netlink_route_multipath_msg_encode(int cmd, struct zebra_dplane_ctx *ctx
 		nl_attr_nest_end(&req->n, nest);
 	}
 
+	/* Count overall nexthops so we can decide whether to use singlepath
+	 * or multipath case.
+	 */
+	nexthop_num = 0;
+	for (ALL_NEXTHOPS_PTR(dplane_ctx_get_ng(ctx), nexthop)) {
+		if (CHECK_FLAG(nexthop->flags, NEXTHOP_FLAG_RECURSIVE))
+			continue;
+		if (!NEXTHOP_IS_ACTIVE(nexthop->flags))
+			continue;
+
+		nexthop_num++;
+	}
+
 	/*
 	 * Always install blackhole routes without using nexthops, because of
 	 * the following kernel problems:
 	 * 1. Kernel nexthops don't suport unreachable/prohibit route types.
 	 * 2. Blackhole kernel nexthops are deleted when loopback is down.
 	 */
 	nexthop = dplane_ctx_get_ng(ctx)->nexthop;
-	if (nexthop) {
+	if (nexthop && nexthop_num == 1) {
```

Comment:
Wouldn't it be better to check the next-hop type instead `NEXTHOP_TYPE_BLACKHOLE`?

Reply:
I mean doing resolved only if not blackhole with `else`.
```c
	nexthop = dplane_ctx_get_ng(ctx)->nexthop;
	if (nexthop) {
		if (nexthop->type == NEXTHOP_TYPE_BLACKHOLE) {
			switch (nexthop->bh_type) {
			case BLACKHOLE_ADMINPROHIB:
				req->r.rtm_type = RTN_PROHIBIT;
				break;
			case BLACKHOLE_REJECT:
				req->r.rtm_type = RTN_UNREACHABLE;
				break;
			case BLACKHOLE_UNSPEC:
			case BLACKHOLE_NULL:
				req->r.rtm_type = RTN_BLACKHOLE;
				break;
			}
			return NLMSG_ALIGN(req->n.nlmsg_len);
		} else {
		        if (CHECK_FLAG(nexthop->flags, NEXTHOP_FLAG_RECURSIVE))
			        nexthop = nexthop->resolved;
                }
	}
```

Reply:
In the case where I encountered the bug, the nexthop was recursively resolved over a blackhole.
```
if (CHECK_FLAG(nexthop->flags, NEXTHOP_FLAG_RECURSIVE))
			        nexthop = nexthop->resolved;
```
It is after this that `nexthop->type == NEXTHOP_TYPE_BLACKHOLE` becomes true.

Reply:
With the suggest change too, zebra would return after processing a nexthop recursively resolved over blackhole, not even looking at the other nexthops in the list. This change tries to avoid that and lets multipath logic take over instead.

Reply:
Sorry, I missed the point that recursively. In any case, could you write down all the details in the commit to avoid other people eyeballing when looking at the history/code?

Reply:
Also, we should have a topotest for such a case.

Reply:
Thanks! Updated the commit message and added a topotest in the latest revision.